pkg/tfvars/tfvars.go

```diff
@@ -21,80 +25,46 @@ type config struct {
 	IgnitionBootstrap string `json:"ignition_bootstrap,omitempty"`
 	IgnitionMaster    string `json:"ignition_master,omitempty"`
 
-	aws.AWS             `json:",inline"`
-	libvirt.Libvirt     `json:",inline"`
-	openstack.OpenStack `json:",inline"`
+	*aws.AWS             `json:",inline"`
+	*libvirt.Libvirt     `json:",inline"`
+	*openstack.OpenStack `json:",inline"`
 }
 
-// TFVars converts the InstallConfig and Ignition content to
-// terraform.tfvar JSON.
-func TFVars(clusterID string, cfg *types.InstallConfig, osImage, bootstrapIgn, masterIgn string) ([]byte, error) {
+// TFVars generates terraform.tfvar JSON for launching the cluster.
+func TFVars(clusterID string, clusterName string, baseDomain string, machineCIDR *net.IPNet, bridge string, region string, externalNetwork string, trunkSupport string, osImage string, bootstrapIgn string, masterIgn string, masters []clusterapi.Machine) ([]byte, error) {
 	config := &config{
-		ClusterID:   clusterID,
-		Name:        cfg.ObjectMeta.Name,
-		BaseDomain:  cfg.BaseDomain,
-		MachineCIDR: cfg.Networking.MachineCIDR.String(),
-
-		IgnitionMaster:    masterIgn,
+		ClusterID:         clusterID,
+		Name:              clusterName,
+		BaseDomain:        baseDomain,
+		MachineCIDR:       machineCIDR.String(),
+		Masters:           len(masters),
 		IgnitionBootstrap: bootstrapIgn,
+		IgnitionMaster:    masterIgn,
 	}
 
-	for _, m := range cfg.Machines {
-		switch m.Name {
-		case "master":
-			var replicas int
-			if m.Replicas == nil {
-				replicas = 1
-			} else {
-				replicas = int(*m.Replicas)
-			}
-
-			config.Masters += replicas
-			if m.Platform.AWS != nil {
-				config.AWS.Master = aws.Master{
-					EC2Type: m.Platform.AWS.InstanceType,
-					MasterRootVolume: aws.MasterRootVolume{
-						IOPS: m.Platform.AWS.EC2RootVolume.IOPS,
-						Size: m.Platform.AWS.EC2RootVolume.Size,
-						Type: m.Platform.AWS.EC2RootVolume.Type,
-					},
-				}
-			}
-		}
+	if config.Masters == 0 {
+		return nil, errors.Errorf("master slice cannot be empty")
 	}
 
-	if cfg.Platform.AWS != nil {
-		config.AWS.Region = cfg.Platform.AWS.Region
-		config.AWS.ExtraTags = cfg.Platform.AWS.UserTags
-		config.AWS.EC2AMIOverride = osImage
-	} else if cfg.Platform.Libvirt != nil {
-		masterIPs := make([]string, len(cfg.Platform.Libvirt.MasterIPs))
-		for i, ip := range cfg.Platform.Libvirt.MasterIPs {
-			masterIPs[i] = ip.String()
-		}
-		config.Libvirt = libvirt.Libvirt{
-			URI: cfg.Platform.Libvirt.URI,
-			Network: libvirt.Network{
-				IfName: cfg.Platform.Libvirt.Network.IfName,
-			},
-			Image:     osImage,
-			MasterIPs: masterIPs,
-		}
-		if err := config.Libvirt.TFVars(&cfg.Networking.MachineCIDR.IPNet, config.Masters); err != nil {
-			return nil, errors.Wrap(err, "failed to insert libvirt variables")
+	var err error
+	machine := masters[0]
+	if awsConfig, ok := machine.Spec.ProviderSpec.Value.Object.(*awsprovider.AWSMachineProviderConfig); ok {
```

Comment:
can we use 
```go
switch provider := machine.Spec.ProviderSpec.Value.Object.(type) {
case *awsprovider.AWSMachineProviderConfig:
...
}
```

Reply:
> can we use...

We could, but then we'd still need casts, wouldn't we?  Seems easier to combine the check with cast in an if/else chain.

Reply:
Hmm... I don't think we need to cast... https://play.golang.org/p/uPq6ZlRtnpk

Reply:
> Hmm... I don't think we need to cast...

With:

```diff
diff --git a/pkg/tfvars/tfvars.go b/pkg/tfvars/tfvars.go
index 1c3c625..5e2cfd6 100644
--- a/pkg/tfvars/tfvars.go
+++ b/pkg/tfvars/tfvars.go
@@ -48,23 +48,25 @@ func TFVars(clusterID string, clusterName string, baseDomain string, machineCIDR
 
 	var err error
 	machine := masters[0]
-	if awsConfig, ok := machine.Spec.ProviderSpec.Value.Object.(*awsprovider.AWSMachineProviderConfig); ok {
-		config.AWS, err = aws.TFVars(awsConfig)
+	machineConfig := machine.Spec.ProviderSpec.Value.Object
+	switch machineConfig.(type) {
+	case *awsprovider.AWSMachineProviderConfig:
+		config.AWS, err = aws.TFVars(machineConfig)
 		if err != nil {
 			return nil, err
 		}
-	} else if libvirtConfig, ok := machine.Spec.ProviderSpec.Value.Object.(*libvirtprovider.LibvirtMachineProviderConfig); ok {
-		config.Libvirt, err = libvirt.TFVars(libvirtConfig, osImage, machineCIDR, bridge, config.Masters)
+	case *libvirtprovider.LibvirtMachineProviderConfig:
+		config.Libvirt, err = libvirt.TFVars(machineConfig, osImage, machineCIDR, bridge, config.Masters)
 		if err != nil {
 			return nil, err
 		}
-	} else if openStackConfig, ok := machine.Spec.ProviderSpec.Value.Object.(*openstackprovider.OpenstackProviderSpec); ok {
-		config.OpenStack, err = openstack.TFVars(openStackConfig, region, externalNetwork, trunkSupport)
+	case *openstackprovider.OpenstackProviderSpec:
+		config.OpenStack, err = openstack.TFVars(machineConfig, region, externalNetwork, trunkSupport)
 		if err != nil {
 			return nil, err
 		}
-	} else {
-		return nil, errors.Errorf("unrecognized master provider %#+v", machine.Spec.ProviderSpec.Value.Object)
+	default:
+		return nil, errors.Errorf("unrecognized master provider %#+v", machineConfig)
 	}
 
 	return json.MarshalIndent(config, "", "  ")
```

on top of 9d94e4204f40, I get:

```console
$ hack/build.sh
...
pkg/tfvars/tfvars.go:54:31: cannot use machineConfig (type runtime.Object) as type *"github.com/openshift/installer/vendor/sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsproviderconfig/v1alpha1".AWSMachineProviderConfig in argument to aws.TFVars: need type assertion
pkg/tfvars/tfvars.go:59:39: cannot use machineConfig (type runtime.Object) as type *"github.com/openshift/installer/vendor/github.com/openshift/cluster-api-provider-libvirt/pkg/apis/libvirtproviderconfig/v1alpha1".LibvirtMachineProviderConfig in argument to libvirt.TFVars: need type assertion
pkg/tfvars/tfvars.go:64:43: cannot use machineConfig (type runtime.Object) as type *"github.com/openshift/installer/vendor/sigs.k8s.io/cluster-api-provider-openstack/pkg/apis/openstackproviderconfig/v1alpha1".OpenstackProviderSpec in argument to openstack.TFVars: need type assertion
```

Am I just doing it wrong?

Reply:
@wking You were missing the part where you store the result of `machineConfig.(type)` and use that value, which will have already been cast to the specific type.